File: codeJava/src/me.imtt/sorting/basic/ShellSort.java
package me.imtt.sorting.basic;

/**
 * 希尔排序（升序）
 */
public class ShellSort {
    public static void sort(Comparable[] arr) {
        int length = arr.length;
        int h = 1;
        while (h < length / 3) {
            h = 3 * h + 1;
        }

        for (; h > 0; h /= 3) {
            for (int i = h; i < length; i++) {
                Comparable temp = arr[i];
                int j = i;
                for (; j >= h && temp.compareTo(arr[j - h]) < 0; j -= h) {
                    arr[j] = arr[j - h];
                }

                arr[j] = temp;
            }
        }
    }
}

File: codePython/01-Sorting-Basic/shell_sort.py
"""
希尔排序（升序）
"""


def sort(arr):
    n = len(arr)
    h = n // 2

    while h > 0:
        for i in range(h, n):
            temp = arr[i]
            j = i
            while j >= h and temp < arr[j - h]:
                arr[j] = arr[j - h]
                j -= h
            arr[j] = temp
        h //= 2
